src/cpp/r/RSexp.cpp

```diff
@@ -838,6 +838,77 @@ bool inherits(SEXP object, const char* S3Class)
    return Rf_inherits(object, S3Class);
 }
 
+std::set<std::string> makeNseFunctions()
+{
+   std::set<std::string> nseFunctions;
+   nseFunctions.insert("quote");
+   nseFunctions.insert("substitute");
+   nseFunctions.insert("match.call");
+   nseFunctions.insert("library");
+   nseFunctions.insert("require");
+   nseFunctions.insert("enquote");
+   nseFunctions.insert("expression");
+   nseFunctions.insert("evalq");
+   nseFunctions.insert("subset");
+   return nseFunctions;
+}
+
+std::set<std::string> nseFunctions()
+{
+   static const std::set<std::string> set = makeNseFunctions();
+   return set;
+}
+
+bool maybePerformsNSEImpl(SEXP node)
+{
+   if (Rf_isExpression(node))
+   {
+      int n = Rf_length(node);
+      for (int i = 0; i < n; i++)
+         if (maybePerformsNSEImpl(VECTOR_ELT(node, i)))
+            return true;
+   }
+   
+   else if (Rf_isLanguage(node))
+   {
+      SEXP symbol = CAR(node);
+      if (Rf_isSymbol(symbol))
```

Comment:
It seems like this should be the base case of the recursion.  For this sort of code I think it's best to use a standard template like http://adv-r.had.co.nz/Expressions.html#ast-funs.  I'd also be hesitant to use any `Rf_is` function because they tend to have surprising behaviour: `TYPEOF(x) == SYMSXP` is clearer